Fang-4-Group/Graduation-Project
# [WIP] [GP-109] Add required package and model implement code

### New Changes
- Add extra-requirements.txt and put torch, torch_geometric in
- Edit Dockerfile, setup.py for extra-requirements.txt
- Add cache pip in my_workflow.yml

@Esther125 @DanielYang1209 @Jeaninetang @frafishen
I have solved the no space left problem, help me chech this pr. Thanks~(●'◡'●)

### Changes
- Add required package:
  - torch, torch_geometric for model
- Add GAT class
- Add model implement

### Test
1. pull this branch
2. run `docker-compose build` because I have add package in requirement
3. run command below to start the server
  ```
  docker-compose up -d
  
  docker exec -it graduation-project /bin/bash
  
  uvicorn src.main:app --host localhost --port 7877 --reload
  ```
5. run the following command to see model result
  ```
  docker exec -it graduation-project /bin/bash
  python src/data_pipeline/model_main.py
  ```
  
  you will see the result below
  ```
  loading .env and geting API_URL variable
  API URL: http://host.docker.internal:7877
  Geting embedding data ...
  Finins geting embedding data
   user id map: {2: 0, 4: 1, 6: 2, 8: 3, 10: 4}
  house id map: {1: 0, 2: 1, 3: 2, 4: 3, 5: 4}
  interactions mapped:
   [[0 0]
   [0 2]
   [1 1]
   [2 4]]
  Epoch 1, Loss: 0.1258
  Epoch 21, Loss: 0.0013
  Epoch 41, Loss: 0.0003
  Epoch 61, Loss: 0.0001
  Epoch 81, Loss: 0.0000
  Epoch 101, Loss: 0.0000
  Epoch 121, Loss: 0.0000
  Epoch 141, Loss: 0.0000
  Epoch 161, Loss: 0.0000
  Epoch 181, Loss: 0.0000
  User Embeddings:
   tensor([[ 1.0737, -0.0558, -1.2378,  1.5170,  1.2500,  1.4207, -2.3968, -0.8989],
          [ 1.4648, -0.9677, -1.6171,  1.5943,  1.8796,  2.8424, -2.3397, -1.0360],
          [ 1.3806,  0.1968, -1.2419,  1.2127,  0.6377,  1.7512, -1.6353, -0.7768],
          [ 1.2495, -0.7861, -2.2129,  1.3614,  0.4992,  2.9479, -1.0917, -0.6241],
          [ 1.6311, -0.7037, -2.1126,  1.1753,  1.0202,  3.5140, -1.9561, -0.3841]])
  Item Embeddings:
   tensor([[ 1.0714, -0.0556, -1.2446,  1.5132,  1.2511,  1.4205, -2.3987, -0.9012],
          [ 1.4681, -0.9691, -1.6135,  1.5991,  1.8796,  2.8410, -2.3397, -1.0338],
          [ 1.0718, -0.0549, -1.2341,  1.5194,  1.2489,  1.4227, -2.3959, -0.8983],
          [ 0.6853, -0.7989, -0.4634,  2.2676,  1.4236,  0.4672, -4.9754, -1.3388],
          [ 1.3822,  0.1975, -1.2417,  1.2078,  0.6375,  1.7515, -1.6338, -0.7779]])
  
  Similarity Matrix:
   [[0.9999976  0.94966865 0.99999905 0.881975   0.95801604]
   [0.9495982  0.999999   0.94963336 0.7595803  0.9430271 ]
   [0.95818764 0.94311774 0.9582697  0.7390937  0.9999988 ]
   [0.82694536 0.92036563 0.82677436 0.53279287 0.90918034]
   [0.87602675 0.9566673  0.87594    0.6122162  0.93715835]]
  
  Top-K Recommendations for each user:
  User 2: [2 0]
  User 4: [1 2]
  User 6: [4 2]
  User 8: [1 4]
  User 10: [1 4]
  ```
